src/ast.rs

```diff
@@ -513,30 +619,65 @@ impl From<(f64, i32)> for NumericConstant {
 pub struct StringConstant(pub String);
 
 // 7. Variable
-// TODO: numeric-array-element is missing
 
-#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+/// Numeric variable containing expressions.
+///
+/// It cannot be stored inside program memory, due to possible deeply nested
+/// expressions in array variables.
+#[derive(Debug, PartialEq)]
 pub enum NumericVariable {
-    Simple { letter: char, digit: Option<u8> },
-    Limit { line_number: u16 },
-    Increment { line_number: u16 },
+    Simple(SimpleNumericVariable),
+    Limit(LimitVariable),
+    Increment(IncrementVariable),
+    Array(ArrayVariable),
 }
 
 impl fmt::Display for NumericVariable {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match self {
-            NumericVariable::Simple { letter, digit } => {
-                f.write_char(*letter)?;
-                if let Some(digit) = digit {
-                    write!(f, "{}", digit)?;
-                }
-                Ok(())
+            NumericVariable::Simple(v) => write!(f, "{}", v),
+            NumericVariable::Array(_) => panic!("logic error: formatting array variable"),
+            NumericVariable::Limit(_) | NumericVariable::Increment(_) => {
+                panic!("logic error: formatting internal variable")
             }
-            _ => panic!("logic error: formatting internal variable"),
         }
     }
 }
 
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+pub enum PlainNumericVariable {
```

Comment:
What is the difference to a `NumericVariable`?

Reply:
There are `ArrayVariables` and `PlainArrayVariables`. The former may contain a numeric expression to evaluate the subscript, the latter contains already evaluated expression, so just a number. As you can see, the former cannot be cloned or hashed, and therefore cannot be stored in the table of all variables, for latter this is perfectly fine.